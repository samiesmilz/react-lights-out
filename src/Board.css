.Board {
  margin: 0 auto;
  align-self: center;
  justify-self: center;
  background-color: rgb(22, 23, 22);
}
