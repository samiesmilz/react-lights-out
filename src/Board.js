import { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows, ncols, chanceLightStartsOn }) {
  const [board, setBoard] = useState(createBoard());

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */

  function createBoard() {
    let initialBoard = [];
    // TODO: create array-of-arrays of true/false values

    for (let i = 0; i < nrows; i++) {
      initialBoard.push(Array(ncols).fill(false));
    }
    // Set some initial values
    initialBoard[1][0] = true;
    initialBoard[1][1] = true;

    return initialBoard;
  }

  function hasWon(board) {
    // Iterate through each cell in the board
    for (let row of board) {
      for (let cell of row) {
        // If any cell is still lit (true), the player hasn't won yet
        if (cell) {
          return false;
        }
      }
    }
    // If all cells are off, the player has won
    return true;
  }

  /** Flip cells around a given cell */
  function flipCellsAround(coord) {
    setBoard((oldBoard) => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      const boardCopy = oldBoard.map((row) => [...row]);

      flipCell(y, x, boardCopy);
      flipCell(y, x - 1, boardCopy);
      flipCell(y, x + 1, boardCopy);
      flipCell(y - 1, x, boardCopy);
      flipCell(y + 1, x, boardCopy);

      return boardCopy;
    });
  }

  // if the game is won, just show a winning msg & render nothing else

  return (
    <>
      <h1>Lets play lights-out</h1>
      <table className="Board">
        <tbody>
          {board.map((row, rowIndex) => (
            <tr key={rowIndex}>
              {row.map((isLit, colIndex) => (
                <Cell
                  key={`${rowIndex}-${colIndex}`}
                  flipCellsAroundMe={() =>
                    flipCellsAround(`${rowIndex}-${colIndex}`)
                  }
                  isLit={isLit}
                />
              ))}
            </tr>
          ))}
        </tbody>
      </table>
      {hasWon(board) && <div>You Win!</div>}
    </>
  );
}

export default Board;
