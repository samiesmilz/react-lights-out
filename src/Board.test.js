import { render, fireEvent } from "@testing-library/react";
import Board from "./Board";

test("renders without errors", () => {
  render(<Board nrows={3} ncols={3} chanceLightStartsOn />);
});

it("matches snapshot...", () => {
  const { asFragment } = render(
    <Board nrows={3} ncols={3} chanceLightStartsOn />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("click event is triggered when cell is clicked", () => {
  const { container } = render(<Board nrows={3} ncols={3} />); // Adjust nrows and ncols accordingly

  // Get the first cell in the table (assuming it's the one you want to test)
  const cell = container.querySelector(".Board tr:first-child td:first-child");
  fireEvent.click(cell);
});
