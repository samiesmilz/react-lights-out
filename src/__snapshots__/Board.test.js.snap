// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot... 1`] = `
<DocumentFragment>
  <h1>
    Lets play lights-out
  </h1>
  <table
    class="Board"
  >
    <tbody>
      <tr>
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
      </tr>
      <tr>
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
